cachelib/navy/common/FdpNvme.cpp

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cachelib/navy/common/FdpNvme.h"
+
+#include <linux/nvme_ioctl.h>
+#include <sys/ioctl.h>
+#include <errno.h>
+#include <cstring>
+#include <numeric>
+
+namespace facebook {
+namespace cachelib {
+namespace navy {
+
+FdpNvme::FdpNvme(int fd) {
+  nvmeData_ = readNvmeInfo(fd);
+  initializeFDP(fd);
+  XLOG(INFO)<< "Creating FdpDevice on fd :" << fd;
+}
+
+int FdpNvme::allocateFdpHandle() {
+  uint16_t phndl;
+
+  // Get NS specific Fdp Placement Handle(PHNDL)
+  if (nextPIDIdx_ <= maxPIDIdx_) {
+    phndl = nextPIDIdx_++;
+  } else {
+    phndl = kDefaultPIDIdx;
+  }
+  // Get Device specific Fdp Placement ID for PHNDL
+  auto pid = getFdpPID(phndl);
+
+  return static_cast<int>(pid);
```

Comment:
Do we need to expose the placement id here? Why not just give out the `phndl`?

Reply:
Yes, possible.